packages/api-client-core/src/types.ts

```diff
@@ -21,15 +21,66 @@ export type LimitToKnownKeys<T, U> = {
  */
 export type VariablesOptions = Record<string, VariableOptions>;
 
+/**
+ * Allows detecting an any type, this is rather tricky:
+ * The type constraint 0 extends 1 is not satisfied (0 is not assignable to 1),
+ * so it should be impossible for 0 extends (1 & T) to be satisfied either, since (1 & T) should be even narrower than 1.
+ * However, when T is any, it reduces 0 extends (1 & any) to 0 extends any, which is satisfied.
+ * That's because any is intentionally unsound and acts as both a supertype and subtype of almost every other type.
+ * source: https://stackoverflow.com/questions/49927523/disallow-call-with-any/49928360#49928360
+ */
+
+type IfAny<T, Y, N> = 0 extends 1 & T ? Y : N;
+
+/**
+ * Convert a schema type into the type that a selection of it must extend
+ *
+ * Example Schema:
+ *
+ * {
+ *   foo: boolean;
+ *   bar?: string;
+ *   nested?: {
+ *     count: number
+ *   }
+ * }
+ *
+ * Example available selection:
+ *
+ * {
+ *   foo?: boolean | null | undefined;
+ *   bar?: boolean | null | undefined;
+ *   nested?: {
+ *     count: boolean | null | undefined
+ *   }
+ * }
+ */
+export type AvailableSelection<Schema> = Schema extends string | number | bigint | null | undefined
+  ? boolean | null | undefined
+  : { [key in keyof Schema]?: AvailableSelection<Schema[key]> };
+
+export type AvailableSelection2<Schema> = Schema extends Array<infer T>
+  ? AvailableSelection2<T>
+  : Schema extends object
+  ? { [key in keyof Schema]?: AvailableSelection2<Schema[key]> }
+  : boolean | null | undefined;
+
 /**
  * Given an options object from a find method, default the type of the selection to a default if no selection is passed
  */
+
 export type DefaultSelection<
   SelectionType,
   Options extends { select?: SelectionType | null },
   Defaults extends SelectionType
 > = Options["select"] extends SelectionType ? Options["select"] : Defaults;
 
+export type DefaultSelection2<
```

Comment:
This one I think I do need for the original PR, tightening the types likely allows the TS inference to do the right thing for me (not 100% sure but at this point I just want to get the whole thing into a working state before tweaking things further). It also fixes it to return the Defaults when the Options ends up resolved to `any` instead of returning `any` which screws everything up.

This is also likely the one that breaks the other packages like `react` because the types are now a lot tighter than before. Again would need to dig more to confirm.